Add missing state and error wrappers

llucax opened the pull request:
- **Remove some unnecessary pylint exceptions**
- **Add a missing `UNKNOWN` EV charger component state**
- **Return `UNKNOWN` instead of `UNSPECIFIED` for unknown states**
- **Wrap missing component errors**
- **Wrap missing component states**
- **Unify the style with the new wrappers**
- **Update release notes**
- **Prepare release notes for the v0.4.0 release**

llucax committed:
Remove some unnecessary pylint exceptions

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add a missing `UNKNOWN` EV charger component state

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Return `UNKNOWN` instead of `UNSPECIFIED` for unknown states

If we can't find the received state between the ones we know, we should
set it to `UNKNOWN`, as there is actually a state, it is **specified**,
but we just don't know about it.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Wrap missing component errors

We add some missing wrappers for battery and inverter errors. To do that
we also need to add new enums for battery and inverter error codes and
the common error levels.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Wrap missing component states

Add some missing wrappers for battery and inverter states and battery
relay states.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Unify the style with the new wrappers

We adjust the documentation and enum lookup to match the newly added
wrappers.

The new enum conversion also avoid a double linear search
when creating enums. We can just try to create it and if it fails
create an `UNSPECIFIED` / `UNKNOWN` item, so we do only one search.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Update release notes

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Prepare release notes for the v0.4.0 release

Adds a summary for the new release.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
I will release v0.4.0 after this is merged.

llucax commented:
Enabled auto-merge.

shsms commented on src/frequenz/client/microgrid/_component_error.py:
```diff
@@ -0,0 +1,198 @@
+        """
+        return cls(
+            code=BatteryErrorCode.from_pb(raw.code),
+            level=ErrorLevel(raw.level),
```
I think we shouldn't assume there will only be `WARN` and `CRITICAL` levels, because we don't know what kind of new component types encounter and what kind of error levels they have.  Much safer to have a `from_pb` for `ErrorLevel` too, like the other enum types.

shsms commented on src/frequenz/client/microgrid/_component_error.py:
```diff
@@ -0,0 +1,198 @@
+        """
+        return cls(
+            code=InverterErrorCode.from_pb(raw.code),
+            level=ErrorLevel(raw.level),
```
here too.

shsms commented on src/frequenz/client/microgrid/_component_states.py:
```diff
@@ -97,3 +192,64 @@ def from_pb(cls, evc_state: ev_charger.ComponentState) -> Self:
+    """
+
+    UNKNOWN = inverter.ComponentState.COMPONENT_STATE_UNKNOWN
+    """The state is provided by the device, but it is not one of the above states."""
```
I think we should say component, not device.

```suggestion
    """A state is provided by the component, but it is not one of the above states."""

llucax replied:
This comes from the API specs, I guess we could change the wording here but not sure if we should _elevate_ it to the API specs. What do you think?

shsms commented on src/frequenz/client/microgrid/_component_states.py:
```diff
@@ -80,6 +80,9 @@ class EVChargerComponentState(Enum):
     """The component is interrupted."""
 
+    UNKNOWN = ev_charger.ComponentState.COMPONENT_STATE_UNKNOWN
+    """The state is provided by the device, but it is not one of the above states."""
```
I think we should say component, not device.

```suggestion
    """A state is provided by the component, but it is not one of the above states."""

shsms commented on src/frequenz/client/microgrid/_component_states.py:
```diff
@@ -6,7 +6,102 @@
+    """The battery is switching off and needs some time to fully shut down."""
+
+    UNKNOWN = battery.ComponentState.COMPONENT_STATE_UNKNOWN
+    """The state is provided by the device, but it is not one of the above states."""
```
I think we should say component, not device.

```suggestion
    """A state is provided by the component, but it is not one of the above states."""

shsms commented on src/frequenz/client/microgrid/_component_states.py:
```diff
@@ -94,6 +94,6 @@ def from_pb(cls, evc_state: ev_charger.ComponentState) -> Self:
         """
         if not any(t.value == evc_state for t in EVChargerComponentState):
-            return cls(cls.UNSPECIFIED)
+            return cls(cls.UNKNOWN)
```
I guess we can't escape having to repurpose something.  Now we are giving extra meaning to `UNKNOWN` instead, saying that either the modbus drivers couldn't recognize the state (and it gets set by the api service), or the client couldn't recognize the state.

shsms replied:
We can have `UNKNOWN_TO_CLIENT = -1` and because protobuf enums can't have negative numbers, it will always be unique, but might be overkill,  and hard to name.

shsms replied:
There are some enums that don't have a `UNKNOWN`, and for those, we're still returning UNSPECIFIED when a code is unknown to the client.  So maybe it is worth making them all consistent.

llucax replied:
Yeah, not sure. If you look at the gRPC errors, some codes can be produced by the server or client, like `INTERNAL` or `RESOURCE_EXHAUSTED` (see https://grpc.github.io/grpc/core/md_doc_statuscodes.html), I think we can see this similarly and say that the estate is unknown, I'm not sure if it is that relevant to the **user of the client** to know if it is unknown to the server or client, I can only see the case where the client is outdated compared to the API specs and the users somehow re-generate its own bindings and use that instead to see if the code is actually known by the server.

IMHO `UNKNOWN` is definitely more accurate here, and if some estate don't have this in the spec, I think that should be a specs bug. @tiyash-basu-frequenz ?

llucax replied:
(ah, Tiyash will probably not answer soon, so we shouldn't wait for his input.

shsms replied:
> IMHO UNKNOWN is definitely more accurate here, and if some estate don't have this in the spec, I think that should be a specs bug.

I think UNKNOWN was introduced for component states where we can have a bug in parsing the raw data.  That's why the the descriptions for all UNKNOWNs say "Component has state, but we understand it".  For other enums where the data doesn't come from hardware, the API doesn't have an UNKNOWN option.

But we are trying to re-purpose UNKNOWN to suggest server-client communication failure, which could happen to all enums, and that's why I think we are trying to give two meanings.

One more option is to say `UNKNOWN = -1` in all python definitions of enums that don't have unknown.

We can keep for later if you prefer, but until then, we'll have UNKNOWN and UNSPECIFIED both being repurposed for something they were not originally meant to do.

llucax replied:
Yeah, I think this goes beyond the microgrid API, maybe it would be a good topic for the API meeting.

llucax committed:
Improve wording for UNKNOWN states

The original wording comes from the API specs docs (for example
https://github.com/frequenz-floss/frequenz-api-microgrid/blob/2332734e86dd8c1d3bf4ee601b93d52525654a6f/proto/frequenz/api/microgrid/inverter.proto#L70-L72),
but we change the wording here to make it more clear.

Co-authored-by: Sahas Subramanian <sahas.subramanian@proton.me>
Signed-off-by: Leandro Lucarella <luca@llucax.com>

llucax commented:
To speed up things I accepted the suggestion to improve the wording for the UNKNOWN estates and added a `from_pb` method to `ErrorLevel`. I disabled auto-merge because I need to squash the fixup once I get a LGTM.

llucax commented:
Actually I'm leaving the auto-merge enabled. Since the fixup commit is not signed-of-by, the automerge won't happen until I squash the commits, so if I get an approval I can squash, force-push and the PR will be queued if the resulting diff doesn't change, so it won't require a second approval after the squashing.

shsms commented:
Fixes look good to me,  can approve once squashed.

shsms reviewed (approved):
oops, just saw you wanted early approval, so you could merge asynchronously.

llucax commented:
You can have another look if you want, but the only that could be messed up is the split of the commits, as long as the resulting diff is the same, the approval still stands.
